server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java

```diff
@@ -43,56 +44,56 @@
+    public CassandraUploadRepository(UploadDAO uploadDAO, BlobStore blobStore, Clock clock) {
         this.uploadDAO = uploadDAO;
         this.blobStore = blobStore;
-        this.bucketNameGenerator = bucketNameGenerator;
         this.clock = clock;
     }
 
     @Override
-    public Publisher<UploadMetaData> upload(InputStream data, ContentType contentType, Username user) {
+    public Mono<UploadMetaData> upload(InputStream data, ContentType contentType, Username user) {
         UploadId uploadId = generateId();
-        UploadBucketName uploadBucketName = bucketNameGenerator.current();
-        BucketName bucketName = uploadBucketName.asBucketName();
 
         return Mono.fromCallable(() -> new CountingInputStream(data))
-            .flatMap(countingInputStream -> Mono.from(blobStore.save(bucketName, countingInputStream, LOW_COST))
-                .map(blobId -> new UploadDAO.UploadRepresentation(uploadId, bucketName, blobId, contentType, countingInputStream.getCount(), user, clock.instant()))
+            .flatMap(countingInputStream -> Mono.from(blobStore.save(UPLOAD_BUCKET, countingInputStream, LOW_COST))
+                .map(blobId -> new UploadDAO.UploadRepresentation(uploadId, blobId, contentType, countingInputStream.getCount(), user, clock.instant()))
                 .flatMap(upload -> uploadDAO.save(upload)
                     .thenReturn(upload.toUploadMetaData())));
     }
 
     @Override
-    public Publisher<Upload> retrieve(UploadId id, Username user) {
+    public Mono<Upload> retrieve(UploadId id, Username user) {
         return uploadDAO.retrieve(user, id)
             .map(upload -> Upload.from(upload.toUploadMetaData(),
-                () -> blobStore.read(upload.getBucketName(), upload.getBlobId(), LOW_COST)))
+                () -> blobStore.read(UPLOAD_BUCKET, upload.getBlobId(), LOW_COST)))
             .switchIfEmpty(Mono.error(() -> new UploadNotFoundException(id)));
     }
 
     @Override
-    public Publisher<Void> delete(UploadId id, Username user) {
+    public Mono<Void> delete(UploadId id, Username user) {
         return uploadDAO.delete(user, id);
     }
 
     @Override
-    public Publisher<UploadMetaData> listUploads(Username user) {
+    public Flux<UploadMetaData> listUploads(Username user) {
         return uploadDAO.list(user)
             .map(UploadDAO.UploadRepresentation::toUploadMetaData);
     }
 
     public Mono<Void> purge() {
-        return Flux.from(blobStore.listBuckets())
-            .<UploadBucketName>handle((bucketName, sink) -> UploadBucketName.ofBucket(bucketName).ifPresentOrElse(sink::next, sink::complete))
-            .filter(bucketNameGenerator.evictionPredicate())
-            .concatMap(bucket -> blobStore.deleteBucket(bucket.asBucketName()))
+        Instant sevenDaysAgo = clock.instant().minus(EXPIRE_DURATION);
+        return Flux.from(uploadDAO.all())
+            .filter(upload -> upload.getUploadDate().isBefore(sevenDaysAgo))
```

Comment:
From what I understood when grooming, I thought the purge was supposed to take out maybe 50% of user's JMAP uploads when running it, purging the oldest uploads based on upload_date? (conclusion being, when hitting the quota limit for the user, we cleanup 50% of its oldest objects in the store)

I hardly see how we would limit quota here if we don't purge anything younger than 7 days... People could still abuse the object store within that 7 days period, exploding their quota, correct? I don't understand how is this logic much different than the current one in place

Reply:
>  I don't understand how is this logic much different than the current one in place

for this task scope, I thought we just changed the mechanism and didn't change the logic.
This method is now invoked by webadmin, I don't know if that is also used when uploading quota
Please correct me

Reply:
The ticket lacks details on that point but from the overall grooming that's how I understood it

I don't see why we should implementing yet an other cleanup task when uploading and reaching quota limit. or should we have some extra parameters? To delete older than 7 days (that seems very arbitrary as a duration by the way) or 50%? 

Maybe @chibenwa can confirm thoughts here

Reply:
Cleanup is less usefull that is for sure.

But if I upload something then go inactive, do I want the docs to remain stored forever?

IMO there is still a use for that cleanup.

Reply:
Good use case, agreed :) Should we have an option to choose between that and 50% then? Or that would come with the ticket applying the upload quota limit when doing an upload via jmap then?

Reply:
IMO both the cleanup and the quota shall apply.